File: spec/challenge.spec.js
'use strict'

// Allow chai syntax like `expect(foo).to.be.ok;`
// jshint -W030

const chai = require('chai')
const chaiAsPromised = require('chai-as-promised')

chai.use(chaiAsPromised)

const expect = chai.expect

const romanNumerals = require('../lib/challenge.js')

describe('Challenge: ', () => {
  describe('arabicToRoman', () => {
    const arabicToRoman = romanNumerals.arabicToRoman

    it('produces simple roman numerals', () => {
      [
        [3, 'III'],
        [20, 'XX'],
        [100, 'C']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(arabicToRoman(arabic)).to.equal(roman)
      })
    })

    it('produces non-subtractive roman numerals', () => {
      [
        [15, 'XV'],
        [6, 'VI'],
        [123, 'CXXIII'],
        [71, 'LXXI']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(arabicToRoman(arabic)).to.equal(roman)
      })
    })

    it('produces subtractive roman numerals', () => {
      [
        [4, 'IV'],
        [9, 'IX'],
        [99, 'XCIX'],
        [405, 'CDV']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(arabicToRoman(arabic)).to.equal(roman)
      })
    })

    it('handles bad input by returning null', () => {
      [0, 0.5, -1, -0.5, '3', null].forEach((badInput) => {
        expect(arabicToRoman(badInput)).to.be.null
      })
    })
  })

  xdescribe('romanToArabic', () => {
    const romanToArabic = romanNumerals.romanToArabic

    it('handles simple roman numerals', () => {
      [
        [3, 'III'],
        [20, 'XX'],
        [100, 'C']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(romanToArabic(roman)).to.equal(arabic)
      })
    })

    it('handles non-subtractive roman numerals', () => {
      [
        [15, 'XV'],
        [6, 'VI'],
        [123, 'CXXIII'],
        [71, 'LXXI']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(romanToArabic(roman)).to.equal(arabic)
      })
    })

    it('handles subtractive roman numerals', () => {
      [
        [4, 'IV'],
        [9, 'IX'],
        [99, 'XCIX'],
        [405, 'CDV']
      ].forEach((pair) => {
        const arabic = pair[0]
        const roman = pair[1]
        expect(romanToArabic(roman)).to.equal(arabic)
      })
    })

    it('handles bad input by returning null', () => {
      ['', 'IIII', 'IM', 'VL', 9, '9', null].forEach((badInput) => {
        expect(romanToArabic(badInput)).to.be.null
      })
    })
  })
})

File: spec/challenge_spec.rb
require_relative '../lib/challenge.rb'
# frozen_string_literal:true

RSpec.describe 'Challenge: ' do
  describe 'arabic_to_roman' do
    it 'produces simple roman numerals' do
      [3, 20, 100].zip(%w[III XX C]).each do |arabic, roman|
        expect(arabic_to_roman(arabic)).to eql(roman)
      end
    end

    it 'produces non-subtractive roman numerals' do
      [15, 6, 123, 71].zip(%w[XV VI CXXIII LXXI]).each do |arabic, roman|
        expect(arabic_to_roman(arabic)).to eql(roman)
      end
    end

    it 'produces subtractive roman numerals' do
      [4, 9, 99, 405].zip(%w[IV IX XCIX CDV]).each do |arabic, roman|
        expect(arabic_to_roman(arabic)).to eql(roman)
      end
    end

    it 'handles bad input by returning nil' do
      [0, 0.5, -1, -0.5, '3', nil].each do |bad_input|
        expect(arabic_to_roman(bad_input)).to be_nil
      end
    end
  end

  xdescribe 'roman_to_arabic' do
    it 'converts simple roman numerals into arabic numbers' do
      [3, 20, 100].zip(%w[III XX C]).each do |arabic, roman|
        expect(roman_to_arabic(roman)).to eql(arabic)
      end
    end

    it 'converts non-subtractive roman numerals into arabic numbers' do
      [15, 6, 123, 71].zip(%w[XV VI CXXIII LXXI]).each do |arabic, roman|
        expect(roman_to_arabic(roman)).to eql(arabic)
      end
    end

    it 'converts subtractive roman numerals into arabic numbers' do
      [4, 9, 99, 405].zip(%w[IV IX XCIX CDV]).each do |arabic, roman|
        expect(roman_to_arabic(roman)).to eql(arabic)
      end
    end

    it 'handles bad input by returning nil' do
      ['', 'IIII', 'IM', 'VL', 9, '9', nil].each do |bad_input|
        expect(roman_to_arabic(bad_input)).to be_nil
      end
    end
  end
end
